irksome/galerkin_stepper.py

```diff
@@ -0,0 +1,274 @@
+    a quadrature rule, produce UFL for the Galerkin-in-Time method.
+
+    :arg F: UFL form for the semidiscrete ODE/DAE
+    :arg L_trial: A :class:`FIAT.FiniteElement` for the trial functions in time
+    :arg L_test: A :class:`FIAT.FinteElement` for the test functions in time
+    :arg Q: A :class:`FIAT.QuadratureRule` for the time integration
+    :arg t: a :class:`Function` on the Real space over the same mesh as
+         `u0`.  This serves as a variable referring to the current time.
+    :arg dt: a :class:`Function` on the Real space over the same mesh as
+         `u0`.  This serves as a variable referring to the current time step.
+         The user may adjust this value between time steps.
+    :arg u0: a :class:`Function` referring to the state of
+         the PDE system at time `t`
+    :arg bcs: optionally, a :class:`DirichletBC` object (or iterable thereof)
+         containing (possibly time-dependent) boundary conditions imposed
+         on the system.
+    :arg nullspace: A list of tuples of the form (index, VSB) where
+         index is an index into the function space associated with `u`
+         and VSB is a :class: `firedrake.VectorSpaceBasis` instance to
+         be passed to a `firedrake.MixedVectorSpaceBasis` over the
+         larger space associated with the Runge-Kutta method
+
+    On output, we return a tuple consisting of four parts:
+
+       - Fnew, the :class:`Form` corresponding to the Galerkin-in-Time discretized problem
+       - UU, the :class:`Function` representing the stages to be solved for
+       - `bcnew`, a list of :class:`firedrake.DirichletBC` objects to be posed
+         on the Galerkin-in-time solution,
+       - 'nspnew', the :class:`firedrake.MixedVectorSpaceBasis` object
+         that represents the nullspace of the coupled system
+    """
+    assert L_test.get_reference_element() == Q.ref_el
+    assert L_trial.get_reference_element() == Q.ref_el
+    assert Q.ref_el.get_spatial_dimension() == 1
+    assert L_trial.get_order() == L_test.get_order() + 1
+
+    v = F.arguments()[0]
+    V = v.function_space()
+    assert V == u0.function_space()
+
+    MC = MeshConstant(V.mesh())
+    vecconst = np.vectorize(MC.Constant)
+
+    num_fields = len(V)
+    num_stages = L_test.space_dimension()
+
+    Vbig = reduce(mul, (V for _ in range(num_stages)))
+    VV = TestFunction(Vbig)
+    UU = Function(Vbig)  # u0 + this are coefficients of the Galerkin polynomial
+
+    qpts = Q.get_points()
+    qwts = Q.get_weights()
+
+    tabulate_trials = L_trial.tabulate(1, qpts)
+    trial_vals = tabulate_trials[(0,)]
+    trial_dvals = tabulate_trials[(1,)]
+    test_vals = L_test.tabulate(0, qpts)[(0,)]
+
+    # sort dofs geometrically by entity location
+    edofs = L_trial.entity_dofs()
+    trial_perm = [*edofs[0][0], *edofs[1][0], *edofs[0][1]]
+    trial_vals = trial_vals[trial_perm]
+    trial_dvals = trial_dvals[trial_perm]
```

Comment:
This should give the right order for point evaluation DOFs, but also handles Bernstein and integral by putting the left/right vertex DOFs as the first and last DOFs.

Reply:
Test space is DG, the interior dofs should be ordered already.

Reply:
FIAT orders the dofs different for continuous and discontinuous Lagrange?

Reply:
I think this is fine, since all DoFs for DG are internal, right?  If so, I guess we're at the mercy of whatever FIAT gives us for their ordering, but probably we shouldn't be second-guessing that without cause